File: 백준/Gold/17298. 오큰수/오큰수.py
N = int(input())
A = list(map(int, input().split()))

ans = [str(-1) for _ in range(N)]
curr = []

for i in range(N-1, -1, -1):
    # if len(curr) == 0:
    #     continue
    if len(curr) != 0:
        if curr[-1] > A[i]:
            ans[i] = str(curr[-1])

    # print(A[i], ans[i], curr, A[i-1])
    if A[i] > A[i-1]:
        curr += [A[i]]
    while len(curr) != 0 and A[i-1] >= curr[-1]:
        curr.pop()
    # print(A[i], ans[i], curr)
    # print()

print(' '.join(ans))

File: 백준/Gold/17298. 오큰수/오큰수.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

		int N = Integer.parseInt(br.readLine());
		int num;

		Stack<Integer> nums = new Stack<>();
		String str = br.readLine();
		String[] strNum = str.split(" ");
		for (int i = 0; i < N; i++) {
			nums.push(Integer.parseInt(strNum[i]));
		}

		Stack<Integer> stk = new Stack<>();
		Stack<Integer> ans = new Stack<>();
		for (int i = 0; i < N; i++) {
			num = nums.pop();

			while (!stk.isEmpty()) {
				if (num >= stk.peek()) {
					stk.pop();
				} else {
					break;
				}
			}

			if (stk.isEmpty()) {
				ans.push(-1);
			} else {
				ans.push(stk.peek());
			}

			stk.push(num);

		}

		while (!ans.isEmpty()) {
			bw.write(ans.pop() + " ");
		}

		bw.flush();
		bw.close();
	}
}
